### PR TITLE
string: add substring method

### PR DESCRIPTION
**Pull request by dabrain34**
This method aims to offer a simple way to 'substring'
an existing string with start and end values.

### CODE REVIEW

File: docs/markdown/Reference-manual.md [hunk 1]
```diff
@@ -1977,6 +1977,9 @@ are immutable, all operations return their results as a new string.
 - `startswith(string)`: returns true if string starts with the string
   specified as the argument
 
+- `substring(start,end)`: returns a substring specified from start to end.
```
Comment (TheQwertiest) [inline docs/markdown/Reference-manual.md:1980]:
```suggestion
- `substring(start,end)` *(since 0.56.0)*: returns a substring specified from start to end.
```

File: docs/markdown/Reference-manual.md [hunk 2]
```diff
@@ -1977,6 +1977,9 @@ are immutable, all operations return their results as a new string.
...
+  Both values are optional as 'foobar'.substring() returns 'foobar'.
```
Comment (TheQwertiest) [inline docs/markdown/Reference-manual.md:1981]:
```suggestion
  Both `start` and `end` arguments are optional, so, for example, `'foobar'.substring()` will return `'foobar'`.
```

File: test cases/common/38 string operations/meson.build [hunk 3]
```diff
@@ -101,3 +101,11 @@ assert('\\\\n' == bs_bs_n, 'Four backslash broken before n')
 assert('\\\\\n' == bs_bs_nl, 'Five backslash broken before n')
 assert('\\\\' == bs_bs, 'Double-backslash broken')
 assert('\\' == bs, 'Backslash broken')
+
+mysubstring='foobarbaz'
+assert(mysubstring.substring() == 'foobarbaz', 'substring is broken')
+assert(mysubstring.substring(0) == 'foobarbaz', 'substring is broken')
+assert(mysubstring.substring(1) == 'oobarbaz', 'substring is broken')
+assert(mysubstring.substring(-5) == 'foobarbaz', 'substring is broken')
```
Comment (tp-m) [inline test cases/common/38 string operations/meson.build:109]: What's the rationale for allowing negative `start` values? Why not just error out?

File: test cases/common/38 string operations/meson.build [hunk 4]
```diff
@@ -101,3 +101,11 @@ assert('\\\\n' == bs_bs_n, 'Four backslash broken before n')
...
+assert(mysubstring.substring(1, 4) == 'oob', 'substring is broken')
+assert(mysubstring.substring(1,-5) == 'oobarbaz', 'substring is broken')
```
Comment (tp-m) [inline test cases/common/38 string operations/meson.build:111]: I would expect an `end` value of `-5` to mean `5 characters from the end`, like python would do for `mysubstring[1:-5]` (which would return `oob` here). Otherwise it should just be a hard error imho.

File: docs/markdown/Syntax.md [hunk 5]
```diff
@@ -220,6 +220,14 @@ is_x86 = target.startswith('x86') # boolean value 'true'
 is_bsd = target.to_lower().endswith('bsd') # boolean value 'true'
 ```
 
+#### .substring()
+
+```meson
+target = 'x86_FreeBSD'
+platform = target.substring(0, 3) # prefix string value 'x86'
+system = target.substring(4) # suffix string value 'FreeBSD'
+```
```
Comment (tp-m) [inline docs/markdown/Syntax.md:229]: Somewhat unfortunate example, because here `split('_')` would arguably be a better fit, but doesn't matter I guess :)

**Comment by xclaesse** [comment]
I think it would be useful to allow negative index like python, to mean relative from the end. 
`'libfoo.so'.substring(3:-3)`would give `foo`.

File: test cases/common/38 string operations/meson.build (hunk 3 above)
Comment (dabrain34) [reply test cases/common/38 string operations/meson.build:109]: because i wanted to avoid errors and I followed javascript sbstring doc (not sure its a good argument :p)

@xclaesse I can implement, so negative end value means pos = len - value right ?

File: docs/markdown/Syntax.md (hunk 5 above)
Comment (dabrain34) [reply docs/markdown/Syntax.md:229]: That's why I implemented this because gstdockingstation.split('gst') is giving you ockination :)

Comment (dabrain34) [reply docs/markdown/Syntax.md:229]: I agree but as a new manner to do it and be sure of of the result :)

File: test cases/common/38 string operations/meson.build (hunk 4 above)
Comment (dabrain34) [reply test cases/common/38 string operations/meson.build]: done

File: mesonbuild/interpreterbase.py [hunk 6]
```diff
@@ -1026,6 +1026,28 @@ def string_method_call(self, obj: str, method_name: str, posargs: T.List[TYPE_nv
...
             return mesonlib.version_compare(obj, cmpr)
+        elif method_name == 'substring':
+            if len(posargs) > 2:
+                raise InterpreterException('substring() takes maximum two arguments.')
+            start = 0
+            end = len(obj)
+            if len (posargs) > 0:
+              start = int(posargs[0])
+            if len (posargs) > 1:
+                end = int(posargs[1])
```
Comment (xclaesse) [inline mesonbuild/interpreterbase.py:1037]: You should check type of args:
```
if not isinstance(posargs[0], int):
    raise InterpreterException()
```

File: mesonbuild/interpreterbase.py [hunk 7]
```diff
@@ -1026,6 +1026,28 @@ def string_method_call(self, obj: str, method_name: str, posargs: T.List[TYPE_nv
...
+            if start < 0:
+                start = 0
```
Comment (xclaesse) [inline mesonbuild/interpreterbase.py:1039]: start could also be relative to end.

File: mesonbuild/interpreterbase.py [hunk 8]
```diff
@@ -1026,6 +1026,28 @@ def string_method_call(self, obj: str, method_name: str, posargs: T.List[TYPE_nv
...
+            if end < 0:
+                if abs(end) <= len(obj):
+                    end = len(obj) - abs(end)
+                else:
+                    end = 0
```
Comment (xclaesse) [inline mesonbuild/interpreterbase.py:1044]: you don't need to implement it yourself, python will do that logic for you already. tbh you don't need to do any checks on start/end values, python does the right thing for any value ;-)

File: mesonbuild/interpreterbase.py [hunk 9]
```diff
@@ -1026,6 +1026,28 @@ def string_method_call(self, obj: str, method_name: str, posargs: T.List[TYPE_nv
...
+            # swap if start > stop
+            if start > end:
+                s = start
+                start = end
+                end = s
```
Comment (xclaesse) [inline mesonbuild/interpreterbase.py:1049]: Why swapping? IMHO `'abc'.substring(1, 0)` should return empty string, like python does.

File: run_unittests.py [hunk 10]
```diff
@@ -2015,6 +2015,10 @@ def test_libdir_must_be_inside_prefix(self):
         else:
             self.assertRaises(subprocess.CalledProcessError, self.setconf, '-Dlibdir=/opt', False)
 
+    def test_string_operations(self):
+      testdir = os.path.join(self.common_test_dir, '38 string operations')
+      self.init(testdir)
```
Comment (xclaesse) [inline run_unittests.py:2020]: you don't need to write unit test here, everything in common/ will be configured/build by default.

Comment (xclaesse) [reply run_unittests.py:2020]: btw, you can just append your tests to `common/38 string operations` where we already test other string methods. It's faster on the CI, it saves meson invocations on hundred small tests.

File: mesonbuild/interpreterbase.py (hunk 9 above)
Comment (dabrain34) [reply mesonbuild/interpreterbase.py]: I took it from javascript as I should have taken from python ;)

File: run_unittests.py (hunk 10 above)
Comment (dabrain34) [reply run_unittests.py]: what is the command to just run this unit test, I was unable to find it.

Comment (xclaesse) [reply run_unittests.py]: I personally have a `~/.local/bin/meson` being a symlink to `meson.py` from source. Then I can just use meson command in test dirs. Dunno if there are better ways.

File: mesonbuild/interpreterbase.py (hunk 9 above)
Comment (xclaesse) [reply mesonbuild/interpreterbase.py]: ah, didn't know JS does that. I guess it depends on the language you're used to... I'm certainly biased toward python, maybe that's not the behaviour others would expect?

File: run_unittests.py (hunk 10 above)
Comment (TheQwertiest) [reply run_unittests.py]: `./meson.py ./run_unit_tests.py test_string_operations`

Comment (dabrain34) [reply run_unittests.py]: yes i removed it at the end ... and thx @xavierclaessens I was dumb as I can run it directly in the folder ... :stuck_out_tongue:

File: test cases/common/38 string operations/meson.build [hunk 11]
```diff
@@ -101,3 +101,11 @@ assert('\\\\n' == bs_bs_n, 'Four backslash broken before n')
...
 assert('\\\\' == bs_bs, 'Double-backslash broken')
 assert('\\' == bs, 'Backslash broken')
+
+mysubstring='foobarbaz'
+assert(mysubstring.substring() == 'foobarbaz', 'substring is broken')
+assert(mysubstring.substring(0) == 'foobarbaz', 'substring is broken')
+assert(mysubstring.substring(1) == 'oobarbaz', 'substring is broken')
+assert(mysubstring.substring(-5) == 'arbaz', 'substring is broken')
+assert(mysubstring.substring(1, 4) == 'oob', 'substring is broken')
+assert(mysubstring.substring(1,-5) == 'oob', 'substring is broken')
```
Comment (xclaesse) [inline test cases/common/38 string operations/meson.build:111]: I think it could also have a few tests with out of range values: (1,0), (0,100), etc.

File: docs/markdown/Syntax.md (hunk 5 above)
Comment (xclaesse) [inline docs/markdown/Syntax.md:229]: Since 0.56

Comment (xclaesse) [reply docs/markdown/Syntax.md:229]: Could also document behaviour of negative and out-of-range values

Comment (xclaesse) [reply docs/markdown/Syntax.md:231]: You forgot that:

> Could also document behaviour of negative and out-of-range values

File: docs/markdown/Syntax.md [hunk 12]
```diff
@@ -220,6 +220,26 @@ is_x86 = target.startswith('x86') # boolean value 'true'
...
+
+```meson
+# Similar to the Python str[start:end] syntax
+target = 'x86_FreeBSD'
+platform = target.substring(0, 3) # prefix string value 'x86'
+system = target.substring(4) # suffix string value 'FreeBSD'
+```
+
+The method accepts negative values where a negative `start` always returns
+empty string and a negative `end` returns from `start` to `len(string) - end`.
```
Comment (xclaesse) [inline docs/markdown/Syntax.md:235]: Negative start also means relative to end, in python `'abc'[-2:-1]` gives `b`.

Comment (dabrain34) [reply docs/markdown/Syntax.md]: I did not find any clear documentation so I was not sure about this behavior, thanks for the clarification. I hope the doc is now clear enough

**Comment by xclaesse** [comment]
Looks all good to me know, thanks! LGTM after release freeze.

File: docs/markdown/Syntax.md [hunk 13]
```diff
@@ -220,6 +220,26 @@ is_x86 = target.startswith('x86') # boolean value 'true'
 is_bsd = target.to_lower().endswith('bsd') # boolean value 'true'
 ```
 
+#### .substring()
+
+Since 0.56.0, you can extract a substring from a string.
```
Comment (TheQwertiest) [inline docs/markdown/Syntax.md:225]:
```suggestion
*(since 0.56.0)*

Extracts a substring from a string.
```

**Comment by xclaesse** [comment]
@dabrain34 could you rebase to see if CI goes green please?

**Comment by xclaesse** [comment]
Since substring() has been proposed by @jpakkane on IRC, I'm assuming it's an approved feature. Merging now, thanks.
